src/relay/transforms/merge_compiler_regions.cc

```diff
@@ -0,0 +1,352 @@
+        // if there is no parent region, move on
+        if (!parent_region.defined()) continue;
+        // merge the parent region if it hasn't been done already
+        if (merged_regions_.find(parent_region->GetID()) == merged_regions_.end()) {
+          VisitExpr(begin->args[0]);
+        }
+        mergeable_regions.push_back(parent_region);
+      }
+      // region_restrictions_[region->id] = std::unordered_set<int>();
+      auto& region_restrictions = region_restrictions_[region->GetID()];
+      for (const auto& parent_region : mergeable_regions) {
+        auto parent_restrictions = region_restrictions_[parent_region->GetID()];
+        // add all the parent restrictions to the current region
+        region_restrictions.insert(parent_restrictions.begin(),
+                                     parent_restrictions.end());
+      }
+      for (const auto& parent_region : mergeable_regions) {
+        bool merged = false;
+        // check the parent region has the same target
+        if (region_targets_[parent_region->GetID()] == compiler_attrs->compiler) {
+          // check the parent region isn't in the restrictions
+          if (region_restrictions.find(parent_region->GetID()) == region_restrictions.end()) {
+            // merge the parent region into the current region
+            regions_->MergeRegions(region, parent_region);
+            // update the restrictions of all other regions to reflect the change in id
+            for (const auto& r : regions_) {
+              auto& restrictions = region_restrictions_[r->GetID()];
+              if (restrictions.find(parent_region->GetID()) != restrictions.end()) {
+                restrictions.erase(parent_region->GetID());
+                restrictions.insert(region->GetID());
+              }
+            }
+            merged = true;
+          }
+        }
+        // if the parent wasn't merged, add it as a restriction to the current region
+        if (!merged)
+          region_restrictions.insert(parent_region->GetID());
+      }
+      merged_regions_.insert(region->GetID());
+    }
+    ExprVisitor::VisitExpr_(call);
+  }
+
+ private:
+  AnnotatedRegionSet regions_;
+  std::unordered_set<int> merged_regions_;
+  std::map<int, std::unordered_set<int>> region_restrictions_;
+  std::map<int, std::string> region_targets_;
+};
+
+
+Expr MergeCompilerRegions(const Expr& expr) {
+  // Annotate the nodes that are not annotated, if any.
+  AnnotateRestDefault anno_default(expr);
+  auto expr_all_annotated = anno_default.Annotate(expr);
+
+  // Create regions using the annotations.
+  AnnotatedRegionSet regions = AnnotatedRegionSet::Create(expr_all_annotated,
+                                                          compiler_begin_op, compiler_end_op);
+
+  // By now, all the nodes have some sort of annotation.
+  // Region merger is ExprVisitor that will update the
+  // AnnotatedRegionSet.
+  RegionMerger merger(regions);
```

Comment:
This name is confusing. `RegionMerger` seems like a mutator but it is actually a visitor. Maybe something like `MergableRegionAnalyzer`. Or it's better to combine this one with `MergeAnnotation`. The logic of `MergeAnnotation` is pretty simple and it can be done in this pass as well, making this pass a real "merger".

Reply:
I don't think I can easily combine the two passes because of the visit order. I agree the name is slightly confusing, but it does merge 'Regions' in that you pass it an AnnotatedRegionSet and it returns a merged version of that set.

Reply:
The regions it merges is not a Relay graph but a separate data structure. It looks weird for a "merger" to only refer the given Relay graph and manipulate a separate data structure.

Another way I can think of to make it more reasonable is moving RegionMerger to the AnnotateRegionSet and make it optional, like:

```
AnnotatedRegionSet regions = AnnotatedRegionSet::Create(
  expr_all_annotated, compiler_begin_op, compiler_end_op, true);
```

where the last argument indicates if the created regions should be merged.

Anyway for the time being, we could revisit this part in the future.